### ISSUE TITLE
ColorInput / ColorPicker issue with React 18

### ISSUE THREAD

**Issue by justinsmid** [mantinedev/mantine#1249]
### What package has an issue

@mantine/core

### Describe the bug

When using NextJS, the ColorInput / ColorPicker components have an issue where their value isn't updated properly when changing the value by clicking on a color in the ColorPicker.

Clicking sets the value to a color that is not the color that was clicked, and the first click sets the value to `null`, as can be seen in the console in the preview below.

I have **not** been able to reproduce it in setups where NextJS is **not** used. In those setups the issue does not occur.

https://user-images.githubusercontent.com/34271675/163398804-b3d28c0e-6443-43d8-93e8-cad96d5c2b20.mp4

### In which browser did the problem occur

I've checked and reproduced it in Chrome and Firefox, have not tried others

### If possible, please include a link to a codesandbox with the reproduced problem

https://codesandbox.io/s/happy-water-51glnv?file=/pages/index.tsx

### Do you know how to fix the issue

No

### Are you willing to participate in fixing this issue and create a pull request with the fix

No

**Comment by rtivital** [comment]
The issue is with React 18 only, downgrade to 17.0.2 and it will work fine.

**Comment by justinsmid** [comment]
Ah, I see.

We'd really like not to downgrade to <18. Is the root cause of the issue known, and/or do you know if the issue lies with Mantine or React? If it's fixable on Mantine's side, do you have plans to look into / fix it?

**Comment by rtivital** [comment]
I do not know how to fix, and I'm not planning to do it until Mantine codebase is migrated to React 18 (maybe a couple of months away). If you want to get the fix quicker, feel free to send a PR.

**Comment by auronsan** [comment]
@rtivital found the fixed but this only takes effect of react 18 caused batching setState

src/mantine-core/src/components/ColorPicker/ColorPicker.tsx:130
v17 need have shouldSkip
```
const handleChange = (color) => {
    setShouldSkip(true);
    setParsed((current) => {
      const next = __spreadValues(__spreadValues({}, current), color);
      valueRef.current = convertHsvaTo(formatRef.current, next);
      return next;
    });
    Promise.resolve().then(() => setValue(valueRef.current)).then(() => setShouldSkip(false));
  };
  ```
  
  but v18 is it unnecessary and make the value not updated on first time.
  ```
const handleChange = (color) => {
    
    setParsed((current) => {
      const next = __spreadValues(__spreadValues({}, current), color);
      valueRef.current = convertHsvaTo(formatRef.current, next);
      return next;
    });
    setValue(valueRef.current);
  };
  ```

**Comment by rtivital** [comment]
Thanks, for investigation, I guess this change can only be implemented with next major version

**Comment by kaloyanBozhkov** [comment]
I ran across this problem today, I can confirm that the first onChange is always null.

the problem is pretty much identified here, hopefully v5 will have this fixed 🤞🏻

**Comment by rtivital** [comment]
Already fixed, will be out with 5.0

**Comment by rtivital** [comment]
Fixed in 5.0
